### PR TITLE
i#6977: Fix quantum underflow

### LINKED ISSUE #6977: scheduler time in quantum underflows, leading to negative values
In testing my per-core-runqueue #6938 implementation I noticed in the logs that the current time in quantum was going negative.  I tracked it down to what looks like both a voluntary switch and preempt at the same time.  There is code that corrects the overshoot for the quantum due to waiting for the next instruction boundary, but its conditions do the wrong thing for the simultaneous triggers.

### PR DESCRIPTION
**Pull request by derekbruening**
Fix bug underflowing quantum time when trying to correct overshoot but the quantum just expired and so is 0; or, we're replaying. Adds an assert.

A test will come in the forthcoming rebalance test for per-output runqueues, which is where this bug was discovered.

Fixes #6977

### CODE REVIEW

**Comment by derekbruening** [comment]
Maybe it is worth filing a separate issue and updating the description to point at it instead of the migrate issue?  Ended up finding and fixing these as part of the runqueues and now trying to split out unrelated pieces with minimal effort...it is probably possible to make a smaller separate dedicated test to trigger this instead of only relying on the new larger rebalance test; tradeoffs.

**Review by brettcoon — APPROVED** [review]
Given the complexity of this code and the number of bugs found recently, it certainly seem like more tests would be beneficial.  But I don't have a good feel for whether or not it's worth adding a small directed test for this particular problem, assuming a bigger test that's already written will reliably cover it.

**Comment by derekbruening** [comment]
Ended up filing an issue to cover this: #6977

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -3555,12 +3557,16 @@ scheduler_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t output,
                 lock.lock();
                 VPRINT(this, 5, "next_record_mid[%d]: switching from %d to %d\n", output,
                        prev_input, outputs_[output].cur_input);
-                if (!preempt) {
+                if (!preempt && // Already reset to 0 for preempt.
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:3560]: The above TODO comment is about the record+instr ordinals on the stream; I updated that comment to clarify.

This is about the time/instrs in the current quantum: I clarified that and which fields it specifically means.

**Comment by derekbruening** [comment]
a64 failure is statecmp-fuzz #6944

**Pull request merged**